test/e2e_federation/hpa.go

```diff
@@ -0,0 +1,605 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	fedclientset "k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset"
+	hpautil "k8s.io/kubernetes/federation/pkg/federation-controller/util/hpa"
+	. "k8s.io/kubernetes/federation/pkg/federation-controller/util/test"
+	"k8s.io/kubernetes/test/e2e/framework"
+	fedframework "k8s.io/kubernetes/test/e2e_federation/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	testHpaPrefix        = "fed-hpa"
+	testHpaTargetObjKind = "Deployment"
+	testDeploymentPrefix = "fed-dep"
+	testHpaScalableImage = "gcr.io/google_containers/hpa-example"
+)
+
+var _ = framework.KubeDescribe("Federated Hpa [Feature:Federation]", func() {
+	f := fedframework.NewDefaultFederatedFramework("federation-hpa")
+
+	Describe("Hpa objects [NoCluster]", func() {
+		var (
+			nsName string
+		)
+
+		BeforeEach(func() {
+			nsName = f.FederationNamespace.Name
+			fedframework.SkipUnlessFederated(f.ClientSet)
+		})
+
+		AfterEach(func() {
+			fedframework.SkipUnlessFederated(f.ClientSet)
+
+			// Delete all hpa objects.
+			nsName := f.FederationNamespace.Name
+			deleteAllHpasOrFail(f.FederationClientset, nsName)
+		})
+
+		It("should be created and deleted successfully", func() {
+			hpa := createHpaOrFail(f.FederationClientset, newHpa(nsName, testHpaPrefix, testDeploymentPrefix, NewInt32(1), NewInt32(50), 4))
+			By(fmt.Sprintf("Creation of hpa %q in namespace %q succeeded.  Deleting hpa.", hpa.Name, nsName))
+			// Cleanup
+			err := f.FederationClientset.Autoscaling().HorizontalPodAutoscalers(nsName).Delete(hpa.Name, &metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "Error deleting hpa %q in namespace %q", hpa.Name, hpa.Namespace)
+			By(fmt.Sprintf("Deletion of hpa %q in namespace %q succeeded.", hpa.Name, nsName))
+		})
+
+		Describe("with valid target obj in spec", func() {
+			AfterEach(func() {
+				// Delete all hpa target objects.
+				deleteAllTargetObjsOrFail(f.FederationClientset, nsName)
+			})
+
+			It("should be able to set correct selected cluster list on target object", func() {
+				// Hpa replicas are distributed randomly; the default distribution can thus
+				// land into any of the clusters. This test verifies against the length of cluster
+				// list. Hpa with min, max = 1, 1 is bound to set the hpa and the target object
+				// only for 1 cluster
+				clusterListLen := 1
+				dep := newHpaTargetObj(nsName, testDeploymentPrefix, NewInt32(0), "")
+				createTargetObjOrFail(f.FederationClientset, dep)
+				hpa := newHpa(nsName, testHpaPrefix, dep.Name, NewInt32(1), NewInt32(50), 1)
+				createHpaOrFail(f.FederationClientset, hpa)
+				waitForClusterListOnTargetObjOrFail(f.FederationClientset, nsName, dep.Name, clusterListLen)
+			})
+		})
+	})
+
+	Describe("Federated Autoscaling: ", func() {
```

Comment:
I would suggest making this consistent with the above Describe clause, to make test filtering easier.  "Federated HPA" for both?